src/map.h

```diff
@@ -260,7 +260,7 @@ class Map
 
 		const Tile* canWalkTo(const Creature& creature, const Position& pos) const;
 
-		bool getPathMatching(const Creature& creature, std::forward_list<Direction>& dirList,
+		bool getPathMatching(const Creature& creature, std::vector<Direction>& dirList,
```

Comment:
I hate this function signature as much as possible. If you are willing, please make it return `std::pair<std::vector<Direction> dirList, bool>` and remove this in-place argument.

Reply:
I don't believe i can change this for performance reasons. It should use a reference to the vector so we could re-use already allocated vectors.

Reply:
But in all usages of these functions, you need to either instantiate a new vector specifically to hold this, or you clear a previously allocated vector which would have the same performance impact. Maybe I missed something here?

Reply:
In few places we directly use the Creature::listWalkDir so no reallocation needed.

Reply:
I see, but the move assignment for vector, which would be triggered when you run `listWalkDir = getPathTo(...)` is exactly the same as running `assign`: it only swaps the pointers.

Reply:
I just checked here:

1) `bool Creature::getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, int32_t minTargetDist, int32_t maxTargetDist, bool fullPathSearch /*= true*/, bool clearSight /*= true*/, int32_t maxSearchDist /*= 0*/) const` is *always* called with a brand new container, so it could return a new vector with the same performance.
2) `bool Creature::getPathTo(const Position& targetPos, std::forward_list<Direction>& dirList, const FindPathParams& fpp) const` sometimes reuse a previous vector.
So maybe you can make 1) work in place, and 2) return a new vector? Not sure if it's worth the hassle though.

Reply:
I got few doubts about this. What is faster? moving the vector or copying a 10-15 byte(steps) memory buffer? This move/copy case mostly happens for player walking and actions code.
My hope for this optimization was to keep the same vector allocated(Creature::listWalkDir) specially for monsters since for every 100 players online there may be 500 or 1000 monsters calculating a new path. That could prevent millions of calls to the allocator which can be expensive(new and and free are in the top 5 tfs most used functions in perf top).

Reply:
Copy elision would kick in here. When you return a vector, the compiler omits copying the returned vector into the assigned vector, performing a move instead (it calls `operator=(const vector&&)`). This is possible because a return value that is not taken as a reference would be immediately destroyed after the return.